src/vcRender.cpp

```diff
@@ -542,15 +543,20 @@ udResult vcRender_AsyncReadFrameDepth(vcRenderContext *pRenderContext)
   return result;
 }
 
-udDouble3 vcRender_DepthToWorldPosition(vcState *pProgramState, vcRenderContext *pRenderContext, double depth)
+udDouble3 vcRender_DepthToWorldPosition(vcState *pProgramState, vcRenderContext *pRenderContext, double depthIn)
 {
-  double nonZeroDepth = (depth == 0.0) ? 1.0 : depth;
+  // reconstruct clip depth from log z
+  float a = pProgramState->settings.camera.farPlane / (pProgramState->settings.camera.farPlane - pProgramState->settings.camera.nearPlane);
+  float b = pProgramState->settings.camera.farPlane * pProgramState->settings.camera.nearPlane / (pProgramState->settings.camera.nearPlane - pProgramState->settings.camera.farPlane);
+  double worldDepth = udPow(2.0, depthIn * udLog2(pProgramState->settings.camera.farPlane + 1.0)) - 1.0;
+  double depth = a + b / worldDepth;
 
   // note: upside down (1.0 - uv.y)
-  udDouble4 clipPos = udDouble4::create(pRenderContext->currentMouseUV.x * 2.0 - 1.0, (1.0 - pRenderContext->currentMouseUV.y) * 2.0 - 1.0, nonZeroDepth, 1.0);
+  udDouble4 clipPos = udDouble4::create(pRenderContext->currentMouseUV.x * 2.0 - 1.0, (1.0 - pRenderContext->currentMouseUV.y) * 2.0 - 1.0, depth, 1.0);
 #if GRAPHICS_API_OPENGL
   clipPos.z = clipPos.z * 2.0 - 1.0;
 #endif
+
   udDouble4 pickPosition4 = pProgramState->camera.matrices.inverseViewProjection * clipPos;
   return pickPosition4.toVector3() / pickPosition4.w;
 }
```

Comment:
Functions like this will need to be wrapped for Metal to work?

Reply:
Metal support will be added later.